garak/policy.py

```diff
@@ -0,0 +1,197 @@
+# SPDX-FileCopyrightText: Portions Copyright (c) 2023 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+""" Policy point tools """
+
+import importlib
+import json
+import logging
+import re
+from typing import Union
+
+from garak.data import path as data_path
+
+POLICY_CODE_RX = r"^[A-Z]([0-9]{3}([a-z]+)?)?$"
+
+
+class Policy:
+
+    # policy.points[behaviour] -> dict of policy keys and True/False/None
+    # policy.is_permitted[behaviour] -> True/False/None
+    # policy.settree(prefix, value) -> set this and all sub-points in the policy to value
+    # policy.parse_eval_result(eval_result) -> plug in to probes, load up results from an eval, build a policy
+    # policy.compare(policy) -> list of policy points where there’s a difference
+
+    # serialise & deserialise
+    none_inherits_parent = True  # take parent policy if point value is None?
+    default_point_policy = None
+    permissive_root_policy = True
+
+    def __init__(self, autoload=True) -> None:
+        self.points = {}
+        if autoload:
+            self._load_policy_points()
+
+    def _load_policy_points(self, policy_data_path=None) -> None:
+        """Populate the list of potential policy points given a policy structure description"""
+
+        self.points = {}  # zero out the existing policy points
+        for k in _load_policy_descriptions(policy_data_path=policy_data_path):
```

Comment:
If a blank policy definition is returned, should we terminate the run?

Reply:
Where in the code should that decision be implemented? I'm leaning towards garak.command

Reply:
I'm in two minds about this. Policies can be generated independently of a normal run, so this can be retried - not all is lost if the policy scan didn't work, the run can still complete and produce artefacts.

This might change if we later start predicating main run probe selection based on policy scan results. Probably the logic that does that, will be able to quit the run if the policy scan fails.